README.md
# Phaser Arcade Slopes Plugin

**Arcade Slopes** brings sloped tile collision handling to
[Phaser](http://phaser.io)'s [Arcade
Physics](http://phaser.io/examples/v2/category/arcade-physics) engine.

## [Demo](http://hexus.github.io/phaser-arcade-slopes)

Check out the **[demo](http://hexus.github.io/phaser-arcade-slopes)**!

![Phaser Arcade Slopes](screenshot.gif)

## Features

- 24 new tile types :tada:
- [SAT](https://github.com/jriecken/sat-js) collision handling :ok_hand:
- Unobtrusive and cooperative integration with Arcade Physics :v:
- Supports sprites :rocket:, groups :busts_in_silhouette:, particle
  emitters :sparkles: and circular physics bodies :white_circle:

## Compatibility

| Phaser Version   | Arcade Slopes Version                                               |
| ---------------- | ------------------------------------------------------------------- |
| v2.4.1 - v2.4.8  | [v0.1.0](https://github.com/hexus/phaser-arcade-slopes/tree/v0.1.0) |
| v2.5.0 - v2.10.1 | [v0.1.1](https://github.com/hexus/phaser-arcade-slopes/tree/v0.1.1) - [v0.3.2](https://github.com/hexus/phaser-arcade-slopes/tree/v0.3.2) |
| v3.x             | [hexus/phaser-slopes](https://github.com/hexus/phaser-slopes)       |

## Installation

Grab a copy of the
[latest release](https://raw.githubusercontent.com/hexus/phaser-arcade-slopes/v0.3.2/dist/phaser-arcade-slopes.min.js)
from the [**dist**](dist) directory in this repository and include it after
Phaser.

```html
<script src="phaser.min.js"></script>
<script src="phaser-arcade-slopes.min.js"></script>
```

## Usage

- [Enabling the plugin](#enabling-the-plugin)
- [Mapping tiles](#mapping-tiles)
- [Enabling physics bodies](#enabling-physics-bodies)
- [Collision](#collision)
- [Debug rendering](#debug-rendering)
- [Extras](#extras)
  - [Minimum Y Offset](#minimum-y-offset)
  - [Collision pulling](#collision-pulling)

### Enabling the plugin

Enable the plugin in the `create()` method of your Phaser state.

```js
game.plugins.add(Phaser.Plugin.ArcadeSlopes);
```

### Mapping tiles

The plugin provides a couple of built in tile slope mappings:

- [Arcade Slopes tileset](assets)
  (`arcadeslopes`)
  ([16px](https://raw.githubusercontent.com/hexus/phaser-arcade-slopes/master/assets/arcade-slopes-16.png),
  [32px](https://raw.githubusercontent.com/hexus/phaser-arcade-slopes/master/assets/arcade-slopes-32.png),
  [64px](https://raw.githubusercontent.com/hexus/phaser-arcade-slopes/master/assets/arcade-slopes-64.png),
  [128px](https://raw.githubusercontent.com/hexus/phaser-arcade-slopes/master/assets/arcade-slopes-128.png))
- [Ninja Physics debug tileset](https://github.com/photonstorm/phaser/tree/v2.4.7/resources/Ninja%20Physics%20Debug%20Tiles)
  (`ninja`)
  ([32px](https://raw.githubusercontent.com/photonstorm/phaser/v2.4.7/resources/Ninja%20Physics%20Debug%20Tiles/32px/ninja-tiles32.png), [64px](https://raw.githubusercontent.com/photonstorm/phaser/v2.4.7/resources/Ninja%20Physics%20Debug%20Tiles/64px/ninja-tiles64.png),
  [128px](https://raw.githubusercontent.com/photonstorm/phaser/v2.4.7/resources/Ninja%20Physics%20Debug%20Tiles/128px/ninja-tiles128.png))

After you've created a tilemap with a collision layer, you'll need to convert
that layer to work with Arcade Slopes.

```js
// Create the tilemap and make it aware of the tileset it uses
map = game.add.tilemap('tilemap');
map.addTilesetImage('collision', 'arcade-slopes-32');
map.setCollisionBetween(1, 38);

// Create the collision layer from the tilemap
ground = map.createLayer('collision');

// Convert the collision layer to work with Arcade Slopes
game.slopes.convertTilemapLayer(ground, 'arcadeslopes');
```

In the case that the first tile ID of the collision tileset in your tilemap is
not `1` (the default), you can provide a third argument to specify it. The
arguments provided to the `setCollisionBetween()` method may need to be adjusted
as well.

```js
map.setCollisionBetween(16, 53);
game.slopes.convertTilemapLayer(ground, 'ninja', 16);
```

**Please note:** Tile GIDs in maps exported from Tiled are always one more than
the ID shown in the GUI. Arcade Slopes expects the above parameter to match the
first GID _as specified in the Tiled map data_.

### Enabling physics bodies

Now you need to enable slopes for any game entities you want to collide against
the tilemap.

```js
game.physics.arcade.enable(player);

game.slopes.enable(player);
game.slopes.enable(emitter);
```

You don't need to do anything special for circular physics bodies, just the
usual `sprite.body.setCircle(radius)`.

_Make sure you call `game.slopes.enable(object)` **after** making any changes to
the **size** or **shape** of the physics body._

### Collision

Now you can collide your sprite against the tilemap in the `update()` method of
your Phaser state, as you normally would, using Arcade Physics. Voila!

```js
// Collide the player with the collision layer
game.physics.arcade.collide(player, ground);

// Collide the particles with the collision layer
game.physics.arcade.collide(emitter, ground);
```

### Debug rendering

To debug your collision layer, set its debug property to `true`.

This will overlay the collision shapes of each tile when the layer is rendered.

```js
ground.debug = true;
```

### Extras

#### Minimum Y offset

This feature separates rectangular physics bodies on the Y axis only, in the
right situations.

```js
// Prefer the minimum Y offset for this physics body
player.body.slopes.preferY = true;

// Prefer the minimum Y offset globally
game.slopes.preferY = true;
```

If you're making a platformer, your player has drag on the X axis, and you don't
want it to slide down slopes, this should solve your problem.

#### Collision pulling

To attempt to keep objects on a surface, you can use collision pulling.

This will pull physics bodies into a collision by a set velocity, if it matches
the set direction.

```js
// Pull the player into downwards collisions with a velocity of 50
player.body.slopes.pullDown = 50;
```

Here are the available properties for collision pulling:

```js
body.slopes.pullUp
body.slopes.pullDown
body.slopes.pullLeft
body.slopes.pullRight
body.slopes.pullTopLeft
body.slopes.pullTopRight
body.slopes.pullBottomLeft
body.slopes.pullBottomRight
```

## Building

If you want to build the plugin yourself from source, clone the repository and
run NPM and Gulp like so.

```bash
npm install
gulp build
```

There's also a watch task that builds the plugin whenever you make changes
to the source.

```bash
gulp watch
```

## Thanks

My thanks go out to those who made this Plugin possible.

- [Richard Davey](https://twitter.com/photonstorm) - for Phaser :rocket:
- [Jim Riecken](https://github.com/jriecken) - [SAT.js](https://github.com/jriecken/sat-js)
  is awesome and saved me loads of time
- [Metanet](http://www.metanetsoftware.com/) - for their incredibly helpful
  tutorials about [collision](http://www.metanetsoftware.com/technique/tutorialA.html)
  [detection](http://www.metanetsoftware.com/technique/tutorialB.html)
- [Olivier Renault](http://elancev.name/oliver/2D%20polygon.htm#tut4) - for their
  tutorial on 2D polygon collision and response (from 2004!)
- [Jan Geselle](https://github.com/geselle-jan) - for writing [a sloped tile
  implementation](https://github.com/geselle-jan/Metroid/commit/9c213e9f5779df1dcd6f7d2bed2a9b676a9e3c6b#diff-467b4e6069f6692511fc5e60f3c426ccR158)
  in Phaser that gave me the idea to write this plugin

CHANGELOG.md
# Arcade Slopes Change Log

## v0.3.2 - 17th August 2021
- Fixed overzealous ignorance of collision normals for less common tile configurations (\#55, \#56)

## v0.3.1 - 1st March 2018
- Implemented Facade.resetCollision() helper method. This can be used to reset
  all Arcade Physics body collision flags to their default values.
- Fixed typescript definition for `game.slopes`.

## v0.3.0 - 5th August 2017
- Removed the snap feature (#34)
- Removed heuristics in favour of a custom SAT implementation that prevents
  internal edge collisions (#36, #38, #40)
- Implemented debug rendering for tile normals and ignored tile normals
- Implemented simple object pooling for less memory-hungry calculations (#42)
- Improved the consistency of SAT responses and overlap values set on physics
  bodies

## v0.2.1 - 1st August 2017
- Dropped bower in favour of npm to load in the SAT.js dependency.
- Added a banner to the minified distributive as part of the build process.
- No functional changes.

## v0.2.0 - 18th June 2017
- Added heuristics for square tiles to improve skipped collisions (#38) at the
  expense of some inconsistencies when bodies exit tiles from inside.
- Allowed physics bodies to disable and enable heuristics for themselves using
  `body.slopes.heuristics`. `null` uses default, `false` disables and `true`
  enables.

## v0.2.0-beta2 - 7th June 2017
- Implemented unintrusive support for an experimental Phaser CE fork.
- Added sprite scale support for AABB physics bodies.
- Included a reference to the plugin with the facade. This makes it easier to
  remove the plugin at runtime
  (`this.game.plugins.remove(this.game.slopes.plugin)`).
- Fixed multiple calls to `slopes.body.enable(sprite)` causing an error.
- Ensured that tilemap debug rendering only wraps if the layer has its `wrap`
  property set to `true`.
- Fixed wasTouching flags (#35).
- Added the last collided tile to `body.slopes.tile`.
- Fixed a heuristics mistake that's been around since May 2016 (4ee0d23c)! There
  are probably more...
- Fixed a wrapping issue with tilemap layer debug rendering.
- Added dynamic `heuristics` and `preferY` properties to the Facade, making it
  simpler to toggle these features globally (`game.slopes.heuristics`,
  `game.slopes.preferY`).
- Fixed SatSolver.collide() setting collision response properties on physics
  bodies before knowing whether the separation was successful.
- Fixed SatSolver.collideOnAxis() not setting collision response properties on
  physics bodies.
- Improved memory consumption by reusing SAT response objects for each body.

## v0.2.0-beta - 10th February 2017
- Supported tile collision flags when determining separation (#27, #28,
  thanks @kevinchau321).
- Fixed collision snapping running when it shouldn't (#21, thanks
  @michaeljcalkins).
- Implemented tile slope polygon debug rendering (#4).
- Added premade Arcade Slopes tilesets (#29).
- Implemented Arcade Slopes tileset mapping shortcut (#9).
- Implemented support for tile property type mapping (#9).

## v0.2.0-alpha2 - 8th February 2017
- Prevented separation if a body isn't moving into a tile (#14).
- Refactored the plugin class to attach the Facade to the Phaser game instead of
  itself.
- Extracted a separate roadmap file from the readme.
- Implemented automatic slope mapping for the Ninja Physics tileset.
- Added missing touching flags for physics bodies (#19, thanks @IkonOne).
- Implemented internal edge flagging (#18).
- Updated SAT.js dependency.
- Implemented support for offset tilemaps (#26).

## v0.2.0-alpha - 19th August 2016
- Initial circular physics body support.
- Typescript definitions (#12, #15, thanks @IkonOne).

## v0.1.1 - 26th June 2016

src/DemoState.js
var DemoState = (function (Phaser) {
	
	/**
	 * @augments {Phaser.State}
	 */
	var DemoState = function () {
		var state = this;
		
		// Feature configuration values that we'll use to control our game state
		this.features = {
			// Arcade slopes
			slopes: true,
			heuristics: false,
			minimumOffsetY: false,
			pullUp: 0,
			pullDown: 0,
			pullLeft: 0,
			pullRight: 0,
			snapUp: 0,
			snapDown: 0,
			snapLeft: 0,
			snapRight: 0,
			
			// Camera controls
			cameraZoom: 1.0,
			cameraMicroZoom: 0.0,
			cameraRotation: 0.0,
			cameraMicroRotation: 0.0,
			cameraLerp: 0.1,
			cameraFollow: true,
			cameraRoundPixels: true,
			cameraSpeed: 20,
			
			// Feature detection
			supports: {
				circleBody: !!Phaser.Physics.Arcade.Body.prototype.setCircle,
			},
			
			// Debug controls
			debugStepMode: function () {
				if (state.game.stepping) {
					state.game.disableStep();
				} else {
					state.game.enableStep();
				}
			},
			debugStep: function () {
				state.game.step();
			},
			debugLayers: true,
			debugLayersFullRedraw: true,
			debugPlayerBody: true,
			debugPlayerBodyInfo: true,
			debugParticleBodies: false,
			debugCameraInfo: false,
			debugInputInfo: false,
			
			// Player controls
			playerHeuristics: false,
			playerMinimumOffsetY: true,
			acceleration: 2000,
			dragX: 1200,
			dragY: 0,
			bounceX: 0,
			bounceY: 0,
			frictionX: 0,
			frictionY: 0,
			jump: 500,
			wallJump: 350,
			shape: 'aabb',
			size: 96,
			anchorX: 0.5,
			anchorY: 0.5,
			
			// Particle controls
			particleFlow: false,
			particleGravity: true,
			particleSelfCollide : false,
			particleMinX: -500,
			particleMaxX: 500,
			particleMinY: -500,
			particleMaxY: 500,
			particleShape: 'circle',
			particleSize: 16,
			particleFrequency: 20,
			particleQuantity: 5,
			emitterWidth: 0.8,
			emitterHeight: 0.8,
			fireParticle: function () {
				state.emitter.emitParticle();
			},
			
			// Tilemaps
			tilemapOffsetX1: 0,
			tilemapOffsetY1: 0,
			tilemapOffsetX2: 0,
			tilemapOffsetY2: 0,
			
			// World
			gravity: 1000,
			enableGravity: true,
			
			// Fun
			slowMotion: 1,
			debug: 0,
		};
	};
	
	DemoState.prototype = {
		
		preload: function () {
			// Load our assets (a demo map and two tilesets)
			this.load.tilemap('demo-tilemap', 'assets/maps/debug.json', null, Phaser.Tilemap.TILED_JSON);
			this.load.spritesheet('arcade-slopes-spritesheet', 'assets/tilesets/arcade-slopes-32-white-8.png', 32, 32);
		},
		
		create: function () {
			var that = this;
			
			// I always have this on :)
			this.time.advancedTiming = true;
			
			// Start up Arcade Physics
			this.physics.startSystem(Phaser.Physics.ARCADE);
			
			// Give it a bit of a boost ;)
			this.game.arcadeSlopesPlugin = this.game.plugins.add(Phaser.Plugin.ArcadeSlopes);
			
			// Set the stage background colour
			this.stage.backgroundColor = '#8d549b';
			
			// Create the tilemap object from the map JSON data
			this.map = this.add.tilemap('demo-tilemap');
			
			// Attach the tileset images to the tilesets defined in the tilemap
			this.map.addTilesetImage('arcade-slopes-32-white-8', 'arcade-slopes-spritesheet');
			
			// Create TilemapLayer objects from the collision layers of the map
			this.ground = this.map.createLayer('collision');
			this.ground2 = this.map.createLayer('collision2');
			this.ground.resizeWorld();
			
			// Enable collision between the appropriate tile indices for each
			// layer in the map
			this.map.setCollisionBetween(1, 38, true, 'collision');
			this.map.setCollisionBetween(1, 38, true, 'collision2');
			
			var newSlopesVersion = compareVersions(Phaser.Plugin.ArcadeSlopes.VERSION, '0.2.0-beta') >= 0;
			var slopeMap = newSlopesVersion ? 'arcadeslopes' : {
				1: 'FULL',
				2: 'HALF_TOP',
				3: 'HALF_BOTTOM',
				4: 'HALF_LEFT',
				5: 'HALF_RIGHT',
				6: 'HALF_BOTTOM_LEFT',
				7: 'HALF_BOTTOM_RIGHT',
				8: 'HALF_TOP_LEFT',
				9: 'HALF_TOP_RIGHT',
				10: 'QUARTER_TOP_LEFT_HIGH',
				11: 'QUARTER_TOP_LEFT_LOW',
				12: 'QUARTER_TOP_RIGHT_LOW',
				13: 'QUARTER_TOP_RIGHT_HIGH',
				14: 'QUARTER_BOTTOM_LEFT_HIGH',
				15: 'QUARTER_BOTTOM_LEFT_LOW',
				16: 'QUARTER_BOTTOM_RIGHT_LOW',
				17: 'QUARTER_BOTTOM_RIGHT_HIGH',
				18: 'QUARTER_LEFT_BOTTOM_HIGH',
				19: 'QUARTER_RIGHT_BOTTOM_HIGH',
				20: 'QUARTER_LEFT_TOP_HIGH',
				21: 'QUARTER_RIGHT_TOP_HIGH',
				35: 'QUARTER_LEFT_BOTTOM_LOW',
				36: 'QUARTER_RIGHT_BOTTOM_LOW',
				37: 'QUARTER_LEFT_TOP_LOW',
				38: 'QUARTER_RIGHT_TOP_LOW'
			};
			
			// Map Arcade Slopes tile types to the correct tilesets, preparing
			// slope data for each tile in the layers
			this.game.slopes.convertTilemapLayer(this.ground, slopeMap);
			this.game.slopes.convertTilemapLayer(this.ground2, slopeMap);
			
			// Create a player sprite
			this.player = this.add.sprite(519, 800);
			this.player.name = 'player';
			
			// Create a graphics object for the player
			this.playerGraphics = new Phaser.Graphics(this.game);
			
			// Generate a texture for the player and give it a physics body
			this.updatePlayer(this.player);
			
			// Set the gravity
			this.physics.arcade.gravity.y = 1000;
			
			// Add a touch of tile padding for the collision detection
			this.player.body.tilePadding.x = 1;
			this.player.body.tilePadding.y = 1;
			
			// Set the initial properties of the player's physics body
			this.player.body.drag.x = this.features.dragX;
			this.player.body.bounce.x = this.features.bounceX;
			this.player.body.bounce.y = this.features.bounceY;
			this.player.body.slopes.friction.x = this.features.frictionX;
			this.player.body.slopes.friction.y = this.features.frictionY;
			this.player.body.maxVelocity.x = 500;
			this.player.body.maxVelocity.y = 1000;
			this.player.body.collideWorldBounds = true;
			
			// Create a particle emitter and position it on the player
			this.emitter = this.add.emitter(this.player.x, this.player.y, 2000);
			
			// Particle graphics
			this.particleGraphics = new Phaser.Graphics(this.game)
				.beginFill(Phaser.Color.hexToRGB('#fff'), 0.5)
				.drawCircle(0, 0, this.features.particleSize);
			
			// Cache the particle graphics as an image
			this.cache.addImage('particle', null, this.particleGraphics.generateTexture().baseTexture.source);
			
			// Create 2000 particles using our newly cached image
			this.emitter.makeParticles('particle', 0, 2000, true, true);
			
			// Give each particle a circular collision body
			if (this.features.supports.circleBody) {
				this.emitter.forEach(function (particle) {
					particle.body.setCircle(Math.floor(that.features.particleSize * 0.5));
				});
			} else {
				this.emitter.forEach(function (particle) {
					particle.body.setSize(that.features.particleSize, that.features.particleSize);
				});
			}
			
			// Attach Arcade Physics polygon data to the particle bodies
			this.game.slopes.enable(this.emitter);
			
			// Set some particle behaviours and properties
			this.emitter.gravity.y = -this.physics.arcade.gravity.y;
			this.emitter.bounce.set(1, 1);
			this.emitter.width = this.player.width;
			this.emitter.height = this.player.height;
			this.emitter.setAlpha(1, 0, 6000);
			this.emitter.setRotation(0, 0);
			this.emitter.setXSpeed(500, 500);
			this.emitter.setYSpeed(0, 0);
			this.emitter._flowTotal = -1;
			
			// Map some keys for use in our update() loop
			this.controls = this.input.keyboard.addKeys({
				'up': Phaser.KeyCode.W,
				'down': Phaser.KeyCode.S,
				'left': Phaser.KeyCode.A,
				'right': Phaser.KeyCode.D,
				'follow': Phaser.KeyCode.F,
				'gravity': Phaser.KeyCode.G,
				'controls': Phaser.KeyCode.C,
				'particles': Phaser.KeyCode.J,
				'toggle': Phaser.KeyCode.K,
				'pause': Phaser.KeyCode.SPACEBAR,
				'cameraUp': Phaser.KeyCode.UP,
				'cameraDown': Phaser.KeyCode.DOWN,
				'cameraLeft': Phaser.KeyCode.LEFT,
				'cameraRight': Phaser.KeyCode.RIGHT
			});
			
			// Follow the player with the camera
			this.camera.follow(this.player);
			
			// Smooth out the camera movement with linear interpolation (lerp)
			this.camera.lerp.setTo(this.features.cameraLerp);
			
			// Register a pointer input event handler that teleports the player
			this.input.onDown.add(function (pointer, mouseEvent) {
				that.player.position.x = pointer.worldX;
				that.player.position.y = pointer.worldY;
				
				// Reset the player's velocity
				that.player.body.velocity.set(0);
			});
			
			// Register a pause handler for Arcade Physics
			this.controls.pause.onDown.add(function () {
				this.physics.arcade.isPaused = !this.physics.arcade.isPaused;
			}, this);
			
			// Register a particle toggle handler
			this.controls.particles.onDown.add(function () {
				that.features.particleFlow = !that.features.particleFlow;
			});
			
			// Prevent the debug text from rendering with a shadow
			this.game.debug.renderShadow = false;
		},
		
		updatePlayer: function (player) {
			var features = this.features;
			var graphics = this.playerGraphics;
			var size = features.size;
			var halfSize = Math.floor(features.size * 0.5);
			
			// Update the player's anchor
			player.anchor.set(features.anchorX, features.anchorY);
			
			// Determine whether we need to update the player
			if (player.body && player.body.height === features.size && ((player.body.isCircle && features.shape === 'circle') || (!player.body.isCircle && features.shape === 'aabb'))) {
				// If the player has a body, and the body's height and shape
				// haven't changed, we don't need to update it
				return;
			}
			
			// Enable physics for the player (give it a physics body)
			this.physics.arcade.enable(player);
			
			// Start the graphics instructions
			graphics.clear();
			graphics._currentBounds = null; // Get Phaser to behave
			graphics.beginFill(Phaser.Color.hexToRGB('#e3cce9'), 1);
			
			// Set a circular physics body
			if (features.shape == 'circle') {
				player.body.setCircle(halfSize);
				graphics.drawCircle(0, 0, features.size);
			}
			
			// Set an AABB physics body
			if (features.shape === 'aabb') {
				player.body.setSize(halfSize, size);
				graphics.drawRect(0, 0, halfSize, size);
			}
			
			// Create a Pixi texture from the graphics and give it to the player
			player.setTexture(graphics.generateTexture(), true);
			
			// We don't have to update the player sprite size, but it's good to
			if (features.shape === 'aabb') {
				player.width = halfSize;
				player.height = size;
			}
			
			if (features.shape === 'circle') {
				player.width = size;
				player.height = size;
			}
			
			// Enable Arcade Slopes physics
			if (this.game.slopes) {
				this.game.slopes.enable(player);
			}
		},
		
		updateEmitter: function (emitter) {
			if (!emitter.children.length) {
				return;
			}
			
			var features = this.features;
			var graphics = this.particleGraphics;
			var size = features.particleSize;
			var halfSize = size / 2;
			var quarterSize = halfSize / 2;
			var firstParticle = emitter.children[0];
			
			// Update the speed range allowed for particles
			this.emitter.setXSpeed(features.particleMinX, features.particleMaxX);
			this.emitter.setYSpeed(features.particleMinY, features.particleMaxY);
			
			// If the particle size option hasn't changed we can finish here
			if (firstParticle.body.height === size && ((firstParticle.body.isCircle && features.shape === 'circle') || (!firstParticle.body.isCircle && features.shape === 'aabb'))) {
				return;
			}
			
			// Draw the particle graphics
			graphics.clear();
			graphics._currentBounds = null; // Get Phaser to behave
			
			graphics.beginFill(Phaser.Color.hexToRGB('#fff'), 0.5)
			
			if (features.particleShape === 'circle') {
				graphics.drawCircle(0, 0, size)
			}
			
			if (features.particleShape === 'aabb') {
				graphics.drawRect(0, 0, size, size)
			}
			
			graphics.updateLocalBounds();
			
			// Cache the graphics as a texture to our particle image cache key
			this.cache.addImage('particle', null, graphics.generateTexture().baseTexture.source);
			
			// Update the body and texture of every particle in the emitter
			emitter.forEach(function (particle) {
				if (features.particleShape === 'circle') {
					particle.body.setCircle(halfSize);
					features.particleShape = 'aabb';
				}
				
				if (features.particleShape === 'aabb') {
					particle.body.setSize(size, size);
				}
				
				particle.loadTexture('particle');
			});
			
			if (this.game.slopes) {
				this.game.slopes.enable(emitter);
			}
		},
		
		update: function () {
			// Define some shortcuts to some useful objects
			var body = this.player.body;
			var camera = this.camera;
			var gravity = this.physics.arcade.gravity;
			var blocked = body.blocked;
			var touching = body.touching;
			var controls = this.controls;
			var features = this.features;
			
			// Phaser feature detection fallbacks
			features.shape = features.shape === 'circle' && !features.supports.circleBody ? 'aabb' : features.shape;
			features.particleShape = features.particleShape === 'circle' && !features.supports.circleBody ? 'aabb' : features.particleShape;
			
			// Update the player
			this.updatePlayer(this.player);
			
			// Update the particle emitter
			this.updateEmitter(this.emitter);
			
			// Update slow motion values; these two are great fun together
			// ( ͡° ͜ʖ ͡°)
			if (this.time.slowMotion !== features.slowMotion) {
				this.time.slowMotion = features.slowMotion;
				this.time.desiredFps = 60 + (features.slowMotion > 1 ? features.slowMotion * 20 : 0);
			}
			
			// Update camera zoom and rotation
			camera.scale.set(
				features.cameraZoom + features.cameraMicroZoom
			);
			camera.rotation = Phaser.Math.degToRad(
				features.cameraRotation + features.cameraMicroRotation
			);
			// this.game.input.scale.set(
			// 	1.0 / (features.cameraZoom + features.cameraMicroZoom)
			// );
			
			// Update camera linear interpolation and pixel rounding
			camera.lerp.set(features.cameraLerp);
			camera.roundPx = features.cameraRoundPixels;
			this.game.renderer.renderSession.roundPixels = features.cameraRoundPixels;
			
			// Toggle camera follow
			if (features.cameraFollow && !camera.target) {
				camera.follow(this.player);
				camera.lerp.set(0.2);
			}
			
			if (!features.cameraFollow && camera.target) {
				camera.unfollow();
			}
			
			// Toggle gravity
			if (controls.gravity.justDown) {
				features.enableGravity = !features.enableGravity;
			}
			
			// Update gravity
			if (features.enableGravity) {
				gravity.y = features.gravity;
			} else {
				gravity.y = 0;
			}
			
			// Update global Arcade Slopes options
			if (this.game.slopes) {
				this.game.slopes.preferY    = features.minimumOffsetY;
				this.game.slopes.heuristics = features.heuristics;
			}
			
			// Update player body properties
			body.drag.x = features.dragX;
			body.drag.y = features.dragY;
			body.bounce.x = features.bounceX;
			body.bounce.y = features.bounceY;
			
			// Update player body Arcade Slopes properties
			body.slopes.friction.x = features.frictionX;
			body.slopes.friction.y = features.frictionY;
			body.slopes.preferY    = features.playerMinimumOffsetY;
			body.slopes.heuristics = features.playerHeuristics;
			body.slopes.pullUp     = features.pullUp;
			body.slopes.pullDown   = features.pullDown;
			body.slopes.pullLeft   = features.pullLeft;
			body.slopes.pullRight  = features.pullRight;
			body.slopes.snapUp     = features.snapUp;
			body.slopes.snapDown   = features.snapDown;
			body.slopes.snapLeft   = features.snapLeft;
			body.slopes.snapRight  = features.snapRight;
			
			// Offset the tilemap layers
			if (this.ground.tileOffset) {
				this.ground.tileOffset.x = features.tilemapOffsetX1;
				this.ground.tileOffset.y = features.tilemapOffsetY1;
				this.ground2.tileOffset.x = features.tilemapOffsetX2;
				this.ground2.tileOffset.y = features.tilemapOffsetY2;
			}
			
			// Debug rendering for the tilemaps
			this.ground.debug = features.debugLayers;
			this.ground2.debug = this.ground.debug;
			
			// Full redraw for the tilemaps
			this.ground.debugSettings.dirty = features.debugLayersFullRedraw;
			this.ground2.debugSettings.dirty = features.debugLayersFullRedraw;
			this.ground.debugSettings.forceFullRedraw = features.debugLayersFullRedraw;
			this.ground2.debugSettings.forceFullRedraw = features.debugLayersFullRedraw;
			
			// Keep the particle emitter attached to the player (though there's
			// probably a better way than this)
			this.emitter.x = this.player.x;
			this.emitter.y = this.player.y;
			this.emitter.width = this.player.width * features.emitterWidth;
			this.emitter.height = this.player.height * features.emitterHeight;
			
			// Update particle lifespan
			this.emitter.lifespan = 3000 / this.time.slowMotion;
			
			// This provides a much better slow motion effect for particles, but
			// because this only affects newly spawned particles, old particles
			// can take ages to die after returning to normal timing
			// TODO: Try updating the lifespan of every particle when slowmotion
			//       changes, in set intervals to avoid it being super slow
			//this.emitter.lifespan = 3000 * this.time.slowMotion;
			//this.emitter.frequency = 1 * this.time.slowMotion;
			//this.emitter.setAlpha(1, 0, 3000 * this.time.slowMotion);
			
			// Ensure that all new particles defy gravity
			if (!features.particleGravity) {
				this.emitter.gravity.y = -this.physics.arcade.gravity.y;
			} else {
				this.emitter.gravity.y = 0;
			}
			
			// Toggle particle flow
			this.emitter.on = !!features.particleFlow;
			this.emitter.lifespan = 3000 / this.time.slowMotion;
			this.emitter.frequency = features.particleFrequency;
			
			if (this.emitter._flowQuantity !== features.particleQuantity) {
				this.emitter._flowQuantity = features.particleQuantity;
				this.emitter._counter = 0;
				this.emitter._timer = this.game.time.time + this.emitter.frequency * this.game.time.slowMotion;
			}
			
			// Toggle the Arcade Slopes plugin itself
			if (features.slopes && !this.game.slopes) {
				this.game.arcadeSlopesPlugin = this.game.plugins.add(Phaser.Plugin.ArcadeSlopes);
			}
			
			if (!features.slopes && this.game.slopes) {
				this.game.plugins.remove(this.game.arcadeSlopesPlugin);
				this.game.arcadeSlopes = null;
			}
			
			// Camera shake for the fun of it
			if (this.input.keyboard.isDown(Phaser.KeyCode.H)) {
				camera.shake(0.005, 50); // :sunglasses:
			}
			
			// Collide the player against the collision layer
			this.physics.arcade.collide(this.player, this.ground);
			this.physics.arcade.collide(this.player, this.ground2);
			
			// Collide the player against the particles
			//this.physics.arcade.collide(this.emitter, this.player);
			
			// Collide the particles against each other
			if (features.particleSelfCollide) {
				this.physics.arcade.collide(this.emitter);
			}
			
			// Collide the particles against the collision layer
			this.physics.arcade.collide(this.emitter, this.ground, function (particle, tile) {
				//console.log(particle.body.slopes);
			});
			
			// Move the camera
			if (controls.cameraUp.isDown) {
				camera.y -= features.cameraSpeed;
			}
			
			if (controls.cameraDown.isDown) {
				camera.y += features.cameraSpeed;
			}
			
			if (controls.cameraLeft.isDown) {
				camera.x -= features.cameraSpeed;
			}
			
			if (controls.cameraRight.isDown) {
				camera.x += features.cameraSpeed;
			}
			
			// Reset the player acceleration
			body.acceleration.x = 0;
			body.acceleration.y = 0;
			
			// Accelerate left
			if (controls.left.isDown) {
				body.acceleration.x = -features.acceleration;
			}
			
			// Accelerate right
			if (controls.right.isDown) {
				body.acceleration.x = features.acceleration;
			}
			
			// Accelerate or jump up
			if (controls.up.isDown) {
				if (features.jump) {
					if (gravity.y > 0 && (blocked.down || touching.down)) {
						body.velocity.y = -features.jump;
					}
				}
				
				if (!features.jump || gravity.y <= 0){
					body.acceleration.y = -Math.abs(gravity.y) - features.acceleration;
				}
			}
			
			// Accelerate down or jump down
			if (controls.down.isDown) {
				if (features.jump) {
					if (gravity.y < 0 && (blocked.up || touching.up)) {
						body.velocity.y = features.jump;
					}
				}
				
				if (!features.jump || gravity.y >= 0){
					body.acceleration.y = Math.abs(gravity.y) + features.acceleration;
				}
			}
			
			// Wall jump
			if (features.wallJump && (controls.up.isDown && gravity.y > 0) || (controls.down.isDown && gravity.y < 0)) {
				if (!(blocked.down || blocked.up || touching.up)) {
					// Would be even better to use collision normals here
					if (blocked.left || touching.left) {
						body.velocity.x = features.wallJump;
						body.velocity.y = gravity.y < 0 ? features.jump : -features.jump;
					}
					
					if (blocked.right || touching.right) {
						body.velocity.x = -features.wallJump;
						body.velocity.y = gravity.y < 0 ? features.jump : -features.jump;
					}
				}
			}
			
			// Stick to ceilings!
			//if (controls.up.isDown && blocked.up) {
			//	body.velocity.y = -50;
			//}
		},
		
		render: function () {
			var debug = this.game.debug;
			var controls = this.controls;
			var features = this.features;
			
			// Render the frame rate
			debug.text(this.time.fps || '--', 4, 16, "#ffffff");
			
			// Render the keyboard controls
			if (controls.controls.isDown) {
				debug.start(32, 196, '#fff', 64);
				debug.line('Click:', 'Teleport');
				debug.line('WASD:', 'Move/jump');
				debug.line('Arrows:', 'Move the camera');
				debug.line('Space:', 'Pause physics');
				debug.line('F:', 'Toggle camera follow');
				debug.line('G:', 'Toggle gravity');
				debug.line('J:', 'Toggle particles');
				debug.line('K:', 'Toggle Arcade Slopes plugin');
				debug.line('C:', 'Show these controls');
				debug.stop();
			}
			
			// Render some debug information about the input, player and camera
			if (features.debugPlayerBody) {
				debug.body(this.player);
			}
			
			if (features.debugPlayerBodyInfo) {
				debug.bodyInfo(this.player, 32, 32);
			}
			
			if (features.debugCameraInfo) {
				debug.cameraInfo(this.camera, 32, 550);
			}
			
			if (features.debugInputInfo) {
				debug.inputInfo(540, 550);
			}
			
			if (features.debugParticleBodies) {
				this.emitter.forEachAlive(function (particle) {
					debug.body(particle);
				});
			}
		}
		
	};
	
	return DemoState;
})(Phaser);
